File: server/src/identity.js
const crypto = require("node:crypto");
const db = require("./db");

const SESSION_COOKIE = "nameplate_session";
// 5 minutes, owner's call: short on purpose. The old model (see schema.sql's
// comment on the retired `devices` table) bound a device to a person forever,
// which kept leaving a shared phone stuck as the wrong person with no
// self-service fix. A short, sliding session means the worst case is "pick
// who you are again" — never "ask someone to go edit the database."
const SESSION_TTL_MS = 5 * 60 * 1000;

const getSession = db.prepare("SELECT user_id AS userId, expires_at AS expiresAt FROM sessions WHERE token = ?");
const insertSession = db.prepare("INSERT INTO sessions (token, user_id, expires_at) VALUES (?, ?, ?)");
const touchSession = db.prepare("UPDATE sessions SET expires_at = ? WHERE token = ?");
const deleteSession = db.prepare("DELETE FROM sessions WHERE token = ?");
const sweepExpired = db.prepare("DELETE FROM sessions WHERE expires_at < ?");
const getUser = db.prepare("SELECT id, name, surname FROM users WHERE id = ?");
const allUsers = db.prepare("SELECT id, name, surname FROM users ORDER BY id");

function cookieOpts() {
  return {
    httpOnly: true,
    sameSite: "lax",
    secure: false, // LAN-only, plain HTTP — see spec §2/§8
    maxAge: SESSION_TTL_MS,
  };
}

/**
 * Resolves the session cookie to a user on every request, and — as long as
 * it's still valid — slides the expiry forward another 5 minutes. That
 * sliding behaviour matters: it means actively using the app (say, working
 * through a round) doesn't log you out mid-task; only the app sitting
 * untouched for a full 5 minutes (or a fresh device with no cookie at all)
 * does. An expired or missing session just means req.session.userId is
 * null — route handlers treat that exactly like "not signed in yet".
 */
function sessionMiddleware(req, res, next) {
  const token = req.cookies ? req.cookies[SESSION_COOKIE] : undefined;
  const session = token ? getSession.get(token) : undefined;

  if (session && session.expiresAt > Date.now()) {
    const expiresAt = Date.now() + SESSION_TTL_MS;
    touchSession.run(expiresAt, token);
    res.cookie(SESSION_COOKIE, token, cookieOpts());
    req.session = { token, userId: session.userId };
  } else {
    if (session) deleteSession.run(token); // stale — don't leave dead rows lying around
    req.session = { token: null, userId: null };
  }
  next();
}

function mePayload(session) {
  const user = session.userId ? getUser.get(session.userId) : null;
  // No "taken" exclusivity any more (see schema.sql) — both names are always
  // pickable, from any device, at any time.
  const candidates = allUsers.all().map((u) => ({ id: u.id, label: u.name }));
  return {
    userId: session.userId,
    label: user ? user.name : null,
    surname: user ? user.surname : null,
    claimed: session.userId != null,
    candidates,
  };
}

/**
 * Starts a fresh 5-minute session for this browser, bound to the chosen
 * person, and sets the cookie for it. Called once per "which one of you is
 * this?" pick — see index.js's POST /api/me.
 */
function selectProfile(res, userId) {
  sweepExpired.run(Date.now()); // opportunistic cleanup, keeps the table small
  const token = crypto.randomUUID();
  insertSession.run(token, userId, Date.now() + SESSION_TTL_MS);
  res.cookie(SESSION_COOKIE, token, cookieOpts());
  return { userId };
}

module.exports = { sessionMiddleware, mePayload, selectProfile, SESSION_COOKIE };

File: server/identity.py
"""Session cookie handling, ported from server/src/identity.js.

CO-4 §3's "session TTL has to change" section: with PINs replacing the old
claim-by-tap model, the reason for a short session (a shared phone stuck as
the wrong person, with no self-service fix) is gone -- you just re-enter your
own PIN. Recommendation taken: 30-day sliding, `secure` on the cookie now
that the site is HTTPS. If this household ever goes back to one shared
phone instead of one-per-parent, shorten SESSION_TTL_MS back down toward 15
minutes so the idle timeout becomes the blind-reveal defence again -- see the
long comment in server/src/identity.js this replaces.

CO-4 §5 drops `candidates` from mePayload -- there is nothing to pick from
any more, the PIN itself is the claim.
"""

import secrets
import time

from transaction import immediate

SESSION_COOKIE = "nameplate_session"
SESSION_TTL_MS = 30 * 24 * 60 * 60 * 1000  # 30 days, sliding. See module docstring.


def _now_ms():
    return int(time.time() * 1000)


def cookie_kwargs():
    return dict(httponly=True, samesite="Lax", secure=True, max_age=SESSION_TTL_MS // 1000)


def resolve_session(conn, token):
    """Returns {"userId": ...} for a still-valid token, else None -- and
    deletes the row if it was present but expired, same opportunistic
    cleanup identity.js did inline."""
    if not token:
        return None
    row = conn.execute(
        "SELECT user_id AS userId, expires_at AS expiresAt FROM sessions WHERE token = ?", (token,)
    ).fetchone()
    if row and row["expiresAt"] > _now_ms():
        return {"userId": row["userId"]}
    if row:
        with immediate(conn):
            conn.execute("DELETE FROM sessions WHERE token = ?", (token,))
    return None


def touch_session(conn, token):
    """Slides the expiry forward. Called once per request that carried a
    valid session, mirroring identity.js's sessionMiddleware."""
    expires_at = _now_ms() + SESSION_TTL_MS
    with immediate(conn):
        conn.execute("UPDATE sessions SET expires_at = ? WHERE token = ?", (expires_at, token))
    return expires_at


def start_session(conn, user_id):
    """Mints a fresh session for a user who just passed the PIN gate
    (gate.py). Ports selectProfile's opportunistic sweep of expired rows."""
    now = _now_ms()
    token = secrets.token_hex(32)
    with immediate(conn):
        conn.execute("DELETE FROM sessions WHERE expires_at < ?", (now,))
        conn.execute(
            "INSERT INTO sessions (token, user_id, expires_at) VALUES (?, ?, ?)",
            (token, user_id, now + SESSION_TTL_MS),
        )
    return token


def me_payload(conn, user_id):
    user = None
    if user_id is not None:
        user = conn.execute("SELECT id, name, surname FROM users WHERE id = ?", (user_id,)).fetchone()
    return {
        "userId": user_id,
        "label": user["name"] if user else None,
        "surname": user["surname"] if user else None,
        "claimed": user_id is not None,
    }
